### PR TITLE
[ECO-4977] Fix broadcast to others

### PR DESCRIPTION
**Pull request by sacOO7**
Related to https://github.com/ably/laravel-broadcaster/issues/48

## Summary by CodeRabbit

- **New Features**
	- Introduced `fromBase64UrlEncoded` and `toBase64UrlEncoded` functions for encoding and decoding Base64 URL strings.
	- Enhanced `socketId` method to return a Base64 encoded JSON object with connection details.

- **Bug Fixes**
	- Improved documentation for the `socketId` method in the Echo class.

- **Tests**
	- Added comprehensive tests for the new utility functions `toBase64UrlEncoded` and `fromBase64UrlEncoded`.
	- Updated test descriptions for clarity regarding public channels.
	- Enhanced mock authentication server with new broadcasting functionality.
	- Updated authorization header construction in tests to use `toBase64UrlEncoded`.

### CODE REVIEW

**Review by ttypic — APPROVED** [review]
Looks good to me, but we probably shouldn't merge until broadcaster part is not ready

**Comment by sacOO7** [comment]
> Looks good to me, but we probably shouldn't merge until broadcaster part is ready

Yes, I am also planning to test current branch in the chat repo. before merging into main.

File: src/connector/ably-connector.ts [hunk 1]
```diff
@@ -118,9 +119,14 @@ export class AblyConnector extends Connector {
...
      * Get the socket ID for the connection.
+     * For ably, returns base64 encoded json with keys {connectionKey, clientId}
      */
     socketId(): string {
-        return this.ably.connection.key;
+        let socketIdObject = {
+            connectionKey : this.ably.connection.key,
+            clientId : this.ably.auth.clientId ?? null,
+        }
+        return toBase64(JSON.stringify(socketIdObject));
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply src/connector/ably-connector.ts]: `socketId` follows standard interface implementation. Please note that, `ablyConnector` class extends `Connector` class. So, we need to maintain `socketId` as a convention.

[the remark this answers is not in the document]
Comment (sacOO7) [reply src/connector/ably-connector.ts]: Also, about backwards compatibility, we have created a separate PR to handle it at a server side -> https://github.com/ably/laravel-broadcaster/pull/50. So, the question is not about compatibility, rather fixing a bug where `socketId` used to return incorrect data.

File: tests/ably/ably-user-rest-publishing.test.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    afterAll(async () => {
+        return await tearDown(testApp);
+    });
+
+    beforeEach(() => {
+        echoInstanes = [];
+    })
+
+    afterEach((done) => {
+        let promises: Array<Promise<Boolean>> = []
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply tests/ably/ably-user-rest-publishing.test.ts]: updated https://github.com/ably-forks/laravel-echo/pull/42/commits/4e78bf60849d16be1eb658521c1c915c2504a1d7

File: tests/ably/ably-user-rest-publishing.test.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+import { AblyChannel, AblyPrivateChannel } from '../../src/channel';
+import * as Ably from 'ably';
+import waitForExpect from 'wait-for-expect';
+import { toText } from '../../src/channel/ably/utils';
+
+jest.setTimeout(30000);
+describe('AblyUserRestPublishing', () => {
+    let testApp: any;
+    let mockAuthServer: MockAuthServer;
+    let echoInstanes: Array<Echo>;
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply tests/ably/ably-user-rest-publishing.test.ts]: updated https://github.com/ably-forks/laravel-echo/pull/42/commits/4e78bf60849d16be1eb658521c1c915c2504a1d7

**Comment by sacOO7** [comment]
@coderabbitai resolve

**Comment by sacOO7** [comment]
@coderabbitai suggest me a string that encodes into base64 string and encoded string has characters + and _.

File: tests/ably/utils.test.ts [hunk 4]
```diff
@@ -1,32 +1,58 @@
...
+            const normalText = "bGFyYXZlbC1lY2hvIGNvZGViYXNlIGlzIG9mIGJlc3QgcXVhbGl0eSwgcGVyaW9kIQ"
+            const encodedText = fromBase64UrlEncoded(normalText);
+            expect(encodedText).toBe('laravel-echo codebase is of best quality, period!')
+
+            // edge cases 
+            expect(fromBase64UrlEncoded('')).toBe('');
+            expect(fromBase64UrlEncoded('SGVsbG8sIOS4lueVjCEg8J-MjQ')).toBe('Hello, 世界! 🌍');
+            expect(fromBase64UrlEncoded('YQ')).toBe('a');  // No padding in Base64Url
+        });
     });
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply tests/ably/utils.test.ts]: updated -> https://github.com/ably-forks/laravel-echo/pull/42/commits/ce52bfb1f1b5d5f50167227f8d5bc68b15d3654b

**Comment by sacOO7** [comment]
@coderabbitai resolve

**Pull request merged**